include/swift/Parse/Parser.h

```diff
@@ -1325,6 +1326,49 @@ class Parser {
   /// Get the location for a type error.
   SourceLoc getTypeErrorLoc() const;
 
+  /// Callback function used for creating a C++ AST from the syntax node at the given source location.
+  ///
+  /// The arguments to this callback are the source file to pass into ASTGen (the exported source file)
+  /// and the source location pointer to pass into ASTGen (to find the syntax node).
+  ///
+  /// The callback returns the new AST node and the ending location of the syntax node. If the AST node
+  /// is NULL, something went wrong.
+  template<typename T>
+  using ASTFromSyntaxTreeCallback = std::pair<T*, const void *>(
+      void *sourceFile, const void *sourceLoc
+  );
+
+  /// Parse by constructing a C++ AST node from the Swift syntax tree via ASTGen.
+  template<typename T>
+  ParserResult<T> parseASTFromSyntaxTree(
+      llvm::function_ref<ASTFromSyntaxTreeCallback<T>> body
+  ) {
+    if (!Context.LangOpts.hasFeature(Feature::ASTGenTypes))
+      return nullptr;
+
+    auto exportedSourceFile = SF.exportedSourceFile;
+    if (!exportedSourceFile)
+      return nullptr;
+
+    // Perform the translation.
+    auto sourceLoc = Tok.getLoc().getOpaquePointerValue();
+    T* astNode;
+    const void *endLocPtr;
+    std::tie(astNode, endLocPtr) = body(exportedSourceFile, sourceLoc);
+
+    if (!astNode) {
+      assert(false && "Could not build AST node from syntax tree");
+      return nullptr;
+    }
+
+    // Spin the lexer until we get to the ending location.
+    while (Tok.getLoc().getOpaquePointerValue() < endLocPtr &&
+           !Tok.is(tok::eof))
+      consumeToken();
```

Comment:
I wonder if it’s faster to advance the lexer by as many bytes as the syntax node takes up instead of consuming tokens. IIRC that’s what I did in my previous ASTGen implementation.

Reply:
Hmm. Advance the lexer and then prime it again; that's better than what I'm doing. Do you have the code lying around?

Reply:
This is the code I used for incremental SwiftSyntax parsing

https://github.com/apple/swift/blob/f3fd46270289b62fa10802f6b5925cbf22290232/lib/Parse/Parser.cpp#L837-L855